File: backend-go/internal/service/auth_service.go
package service

import (
	"time"

	"github.com/leavemarker/backend-go/internal/dto"
	"github.com/leavemarker/backend-go/internal/enums"
	"github.com/leavemarker/backend-go/internal/middleware"
	"github.com/leavemarker/backend-go/internal/models"
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

// AuthService handles authentication logic
type AuthService struct {
	db               *gorm.DB
	companyRepo      *repository.CompanyRepository
	employeeRepo     *repository.EmployeeRepository
	planRepo         *repository.PlanRepository
	subscriptionRepo *repository.SubscriptionRepository
	jwtService       *middleware.JWTService
	emailService     *EmailService
}

// NewAuthService creates a new AuthService
func NewAuthService(
	db *gorm.DB,
	companyRepo *repository.CompanyRepository,
	employeeRepo *repository.EmployeeRepository,
	planRepo *repository.PlanRepository,
	subscriptionRepo *repository.SubscriptionRepository,
	jwtService *middleware.JWTService,
	emailService *EmailService,
) *AuthService {
	return &AuthService{
		db:               db,
		companyRepo:      companyRepo,
		employeeRepo:     employeeRepo,
		planRepo:         planRepo,
		subscriptionRepo: subscriptionRepo,
		jwtService:       jwtService,
		emailService:     emailService,
	}
}

// Signup registers a new company and admin user
func (s *AuthService) Signup(req dto.SignupRequest) (*dto.JWTAuthResponse, error) {
	// Check if company email exists
	exists, err := s.companyRepo.ExistsByEmail(req.CompanyEmail)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to check company email")
	}
	if exists {
		return nil, utils.NewBadRequestError("Company email already exists")
	}

	// Check if employee email exists
	exists, err = s.employeeRepo.ExistsByEmail(req.Email)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to check employee email")
	}
	if exists {
		return nil, utils.NewBadRequestError("Employee email already exists")
	}

	// Hash password
	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to hash password")
	}

	var employee *models.Employee

	// Start transaction
	err = s.db.Transaction(func(tx *gorm.DB) error {
		// Create company
		company := &models.Company{
			Name:     req.CompanyName,
			Email:    req.CompanyEmail,
			Timezone: "Asia/Kolkata",
			Active:   true,
		}
		if err := tx.Create(company).Error; err != nil {
			return err
		}

		// Create admin employee
		employee = &models.Employee{
			CompanyID:      company.ID,
			EmployeeID:     req.EmployeeID,
			FullName:       req.FullName,
			Email:          req.Email,
			Password:       hashedPassword,
			Role:           enums.RoleSuperAdmin,
			WorkLocation:   req.WorkLocation,
			EmploymentType: enums.EmploymentTypeFullTime,
			Status:         enums.EmployeeStatusActive,
		}
		if err := tx.Create(employee).Error; err != nil {
			return err
		}

		// Create free subscription
		freePlan, err := s.planRepo.FindFreePlan()
		if err != nil {
			// Create default free plan if not exists
			freePlan = &models.Plan{
				Name:             "Free",
				Description:      "Free tier plan",
				Tier:             enums.PlanTierFree,
				PlanType:         enums.PlanTypeFree,
				MaxEmployees:     10,
				MaxLeavePolicies: 1,
				MaxHolidays:      6,
				Active:           true,
			}
			if err := tx.Create(freePlan).Error; err != nil {
				return err
			}
		}

		now := time.Now()
		subscription := &models.Subscription{
			CompanyID:          company.ID,
			PlanID:             freePlan.ID,
			Status:             enums.SubscriptionStatusActive,
			BillingCycle:       enums.BillingCycleMonthly,
			StartDate:          now,
			EndDate:            now.AddDate(100, 0, 0), // Free plan doesn't expire
			CurrentPeriodStart: now,
			CurrentPeriodEnd:   now.AddDate(100, 0, 0),
			Amount:             decimal.Zero,
			AutoRenew:          false,
			IsPaid:             true,
		}
		if err := tx.Create(subscription).Error; err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		return nil, utils.NewInternalServerError("Failed to create account")
	}

	// Generate JWT token
	token, err := s.jwtService.GenerateToken(employee.ID, employee.Email, employee.FullName, employee.Role, employee.CompanyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to generate token")
	}

	return &dto.JWTAuthResponse{
		AccessToken: token,
		TokenType:   "Bearer",
		UserID:      employee.ID,
		Email:       employee.Email,
		FullName:    employee.FullName,
		Role:        employee.Role,
		CompanyID:   employee.CompanyID,
	}, nil
}

// Login authenticates a user
func (s *AuthService) Login(req dto.LoginRequest) (*dto.JWTAuthResponse, error) {
	employee, err := s.employeeRepo.FindByEmail(req.Email)
	if err != nil {
		return nil, utils.NewUnauthorizedError("Invalid email or password")
	}

	if employee.Status != enums.EmployeeStatusActive {
		return nil, utils.NewUnauthorizedError("Account is inactive")
	}

	if !utils.CheckPassword(req.Password, employee.Password) {
		return nil, utils.NewUnauthorizedError("Invalid email or password")
	}

	// Generate JWT token
	token, err := s.jwtService.GenerateToken(employee.ID, employee.Email, employee.FullName, employee.Role, employee.CompanyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to generate token")
	}

	return &dto.JWTAuthResponse{
		AccessToken: token,
		TokenType:   "Bearer",
		UserID:      employee.ID,
		Email:       employee.Email,
		FullName:    employee.FullName,
		Role:        employee.Role,
		CompanyID:   employee.CompanyID,
	}, nil
}

// VerifySession verifies a user session
func (s *AuthService) VerifySession(userID uint) (*dto.UserSessionResponse, error) {
	employee, err := s.employeeRepo.FindByID(userID)
	if err != nil {
		return nil, utils.NewUnauthorizedError("Session invalid")
	}

	if employee.Status != enums.EmployeeStatusActive {
		return nil, utils.NewUnauthorizedError("Account is inactive")
	}

	return &dto.UserSessionResponse{
		ID:        employee.ID,
		Email:     employee.Email,
		FullName:  employee.FullName,
		Role:      employee.Role,
		CompanyID: employee.CompanyID,
	}, nil
}

// RequestPasswordReset initiates password reset
func (s *AuthService) RequestPasswordReset(req dto.PasswordResetRequest) error {
	employee, err := s.employeeRepo.FindByEmail(req.Email)
	if err != nil {
		// Don't reveal if email exists
		return nil
	}

	// Generate reset token
	token, err := utils.GenerateRandomToken(32)
	if err != nil {
		return utils.NewInternalServerError("Failed to generate reset token")
	}

	// Set token and expiry (24 hours)
	expiry := time.Now().Add(24 * time.Hour)
	employee.PasswordResetToken = &token
	employee.PasswordResetTokenExpiry = &expiry

	if err := s.employeeRepo.Update(employee); err != nil {
		return utils.NewInternalServerError("Failed to save reset token")
	}

	// Send email (ignore error for security)
	_ = s.emailService.SendPasswordResetEmail(employee.Email, token, employee.FullName)

	return nil
}

// ResetPassword resets password with token
func (s *AuthService) ResetPassword(req dto.PasswordResetConfirmRequest) error {
	employee, err := s.employeeRepo.FindByPasswordResetToken(req.Token)
	if err != nil {
		return utils.NewBadRequestError("Invalid or expired reset token")
	}

	// Check if token is expired
	if employee.PasswordResetTokenExpiry == nil || time.Now().After(*employee.PasswordResetTokenExpiry) {
		return utils.NewBadRequestError("Reset token has expired")
	}

	// Hash new password
	hashedPassword, err := utils.HashPassword(req.NewPassword)
	if err != nil {
		return utils.NewInternalServerError("Failed to hash password")
	}

	// Update password and clear token
	employee.Password = hashedPassword
	employee.PasswordResetToken = nil
	employee.PasswordResetTokenExpiry = nil

	if err := s.employeeRepo.Update(employee); err != nil {
		return utils.NewInternalServerError("Failed to update password")
	}

	return nil
}

File: backend/src/main/java/com/leavemarker/service/AuthService.java
package com.leavemarker.service;

import com.leavemarker.dto.auth.*;
import com.leavemarker.entity.Company;
import com.leavemarker.entity.Employee;
import com.leavemarker.entity.Plan;
import com.leavemarker.entity.Subscription;
import com.leavemarker.enums.BillingCycle;
import com.leavemarker.enums.EmployeeStatus;
import com.leavemarker.enums.EmploymentType;
import com.leavemarker.enums.PlanTier;
import com.leavemarker.enums.Role;
import com.leavemarker.enums.SubscriptionStatus;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.exception.ResourceNotFoundException;
import com.leavemarker.repository.CompanyRepository;
import com.leavemarker.repository.EmployeeRepository;
import com.leavemarker.repository.PlanRepository;
import com.leavemarker.repository.SubscriptionRepository;
import com.leavemarker.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final EmployeeRepository employeeRepository;
    private final CompanyRepository companyRepository;
    private final PlanRepository planRepository;
    private final SubscriptionRepository subscriptionRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;

    @Transactional
    public JwtAuthResponse signup(SignupRequest request) {
        if (companyRepository.existsByEmailAndDeletedFalse(request.getCompanyEmail())) {
            throw new BadRequestException("Company email already exists");
        }

        if (employeeRepository.existsByCompanyIdAndEmailAndDeletedFalse(null, request.getEmail())) {
            throw new BadRequestException("Employee email already exists");
        }

        Company company = Company.builder()
                .name(request.getCompanyName())
                .email(request.getCompanyEmail())
                .timezone("Asia/Kolkata")
                .active(true)
                .build();
        company = companyRepository.save(company);

        Employee employee = Employee.builder()
                .company(company)
                .employeeId(request.getEmployeeId())
                .fullName(request.getFullName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.SUPER_ADMIN)
                .dateOfJoining(LocalDate.now())
                .employmentType(EmploymentType.FULL_TIME)
                .workLocation(request.getWorkLocation())
                .status(EmployeeStatus.ACTIVE)
                .build();
        employee = employeeRepository.save(employee);

        // Auto-create FREE subscription for new companies
        createFreeSubscription(company);

        String jwt = tokenProvider.generateTokenFromEmail(
                employee.getEmail(),
                employee.getId(),
                employee.getRole().name(),
                company.getId()
        );

        return JwtAuthResponse.builder()
                .accessToken(jwt)
                .userId(employee.getId())
                .email(employee.getEmail())
                .fullName(employee.getFullName())
                .role(employee.getRole().name())
                .companyId(company.getId())
                .build();
    }

    public JwtAuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);
        String jwt = tokenProvider.generateToken(authentication);

        Employee employee = employeeRepository.findByEmailAndDeletedFalse(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found"));

        return JwtAuthResponse.builder()
                .accessToken(jwt)
                .userId(employee.getId())
                .email(employee.getEmail())
                .fullName(employee.getFullName())
                .role(employee.getRole().name())
                .companyId(employee.getCompany().getId())
                .build();
    }

    @Transactional
    public void requestPasswordReset(PasswordResetRequest request) {
        Employee employee = employeeRepository.findByEmailAndDeletedFalse(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found with email: " + request.getEmail()));

        String resetToken = UUID.randomUUID().toString();
        employee.setPasswordResetToken(resetToken);
        employee.setPasswordResetTokenExpiry(LocalDate.now().plusDays(1));
        employeeRepository.save(employee);

        // TODO: Send email with reset token
        // For now, we'll just log it
        System.out.println("Password reset token: " + resetToken);
    }

    @Transactional
    public void resetPassword(PasswordResetConfirmRequest request) {
        Employee employee = employeeRepository.findByPasswordResetTokenAndDeletedFalse(request.getToken())
                .orElseThrow(() -> new BadRequestException("Invalid or expired reset token"));

        if (employee.getPasswordResetTokenExpiry().isBefore(LocalDate.now())) {
            throw new BadRequestException("Reset token has expired");
        }

        employee.setPassword(passwordEncoder.encode(request.getNewPassword()));
        employee.setPasswordResetToken(null);
        employee.setPasswordResetTokenExpiry(null);
        employeeRepository.save(employee);
    }

    /**
     * Creates a FREE subscription for a newly registered company.
     * The FREE plan allows up to 10 employees with basic features.
     */
    private void createFreeSubscription(Company company) {
        Plan freePlan = planRepository.findFirstByTierAndActiveTrueAndDeletedFalse(PlanTier.FREE)
                .orElse(null);

        if (freePlan == null) {
            log.warn("No active FREE plan found. Company {} will not have a subscription.", company.getId());
            return;
        }

        LocalDateTime now = LocalDateTime.now();
        // FREE plan never expires - set to far future
        LocalDateTime farFuture = now.plusYears(100);

        Subscription subscription = Subscription.builder()
                .company(company)
                .plan(freePlan)
                .status(SubscriptionStatus.ACTIVE)
                .billingCycle(BillingCycle.MONTHLY)
                .startDate(now)
                .endDate(farFuture)
                .currentPeriodStart(now)
                .currentPeriodEnd(farFuture)
                .amount(BigDecimal.ZERO)
                .autoRenew(false)
                .isPaid(true) // FREE plan is always "paid"
                .build();

        subscriptionRepository.save(subscription);
        log.info("Created FREE subscription for company: {}", company.getName());
    }
}
